c10/core/TensorTypeSet.h

```diff
@@ -0,0 +1,84 @@
+#pragma once
+
+#include <c10/core/TensorTypeId.h>
+#include <c10/util/llvmMathExtras.h>
+#include <c10/util/Exception.h>
+#include <ostream>
+
+namespace c10 {
+
+// A representation of a set of TensorTypeIds.  A tensor may
+// have multiple tensor type ids, e.g., a Variable tensor
+// can also be a CPU tensor; the TensorTypeSet specifies what
+// type ids apply.  The internal representation is as a 64-bit
+// bit set (this means only 64 tensor type ids are supported).
+//
+// At the moment, there are no nontrivial uses of this set; tensors
+// are always singletons.  In the near future, this set will represent
+// variable? + tensor type id.  In the far future, it will be
+// variable? + profiling? + tracing? + lazy? + tensor type id.
+//
+// An undefined tensor is one with an empty tensor type set.
+class TensorTypeSet {
+public:
+  TensorTypeSet() {}
+  explicit TensorTypeSet(TensorTypeId t)
+    : repr_(t == TensorTypeId::UndefinedTensorId
+              ? 0
+              : 1ULL << (static_cast<uint8_t>(t) - 1)) {}
+  // Test if a TensorTypeId is in the set
+  bool has(TensorTypeId t) const {
+    TORCH_INTERNAL_ASSERT(t != TensorTypeId::UndefinedTensorId);
+    return static_cast<bool>(repr_ & TensorTypeSet(t).repr_);
+  }
+  // Perform set union
+  TensorTypeSet operator|(TensorTypeSet other) const {
+    return TensorTypeSet(repr_ | other.repr_);
+  }
+  // Perform set equality
+  bool operator==(TensorTypeSet other) const {
+    return repr_ == other.repr_;
+  }
+  // Add a TensorTypeId to the TensorTypeId set.  Does NOT mutate,
+  // returns the extended TensorTypeSet!
+  C10_NODISCARD TensorTypeSet add(TensorTypeId t) const {
+    return *this | TensorTypeSet(t);
+  }
+  // Remove a TensorTypeId from the TensorTypeId set.  This is
+  // generally not an operation you should be doing (it's
+  // used to implement operator<<)
+  C10_NODISCARD TensorTypeSet remove(TensorTypeId t) const {
+    return TensorTypeSet(repr_ & ~TensorTypeSet(t).repr_);
+  }
+  // Is the set empty?  (AKA undefined tensor)
+  bool empty() const {
+    return repr_ == 0;
+  }
+  // Return the "first" type id in this set; i.e., the one that
+  // should handle dispatch for this operator
+  TensorTypeId firstTypeId() const {
```

Comment:
this is kind of scary from the API perspective (maybe you get rid of this later?).  But clearly the order of TensorTypeId matters, but that's totally opaque here.  Why is this different from dispatchTypeId?

Reply:
> maybe you get rid of this later?

No, it always exists as piece of API on TensorTypeSet.

> But clearly the order of TensorTypeId matters, but that's totally opaque here.

I can add this to the comment.

> Why is this different from dispatchTypeId?

firstTypeId is a pure function on the bitset: it's always well-defined what the first bit in the bitset is. However, dispatchTypeId must account for TLS when determining what the actual dispatch should be. So that lives external to this class. (Underlying design principle: if you write a data structure, put only pure functions on the data structure, and let external code handle interaction with the external world, e.g., TLS)

Reply:
Renaming this to highestPriorityTypeId